torch/_dynamo/guards.py

```diff
@@ -737,45 +776,90 @@ def compile_check_fn(
-        verbose_code_parts.extend(local_builder.shape_env_code)
         assert not global_builder.shape_env_code
 
-        code = " and ".join(unique(code_parts))
         closure_vars = collections.OrderedDict(
             [
                 ("___guarded_code", self),
                 ("___check_tensors", check_tensors_fn),
-                ("___check_tensors_verbose", check_tensors_verbose_fn),
+                ("___check_tensors_verbose", check_tensor_verbose),
                 ("tensor_check_names", tensor_check_names),
+                ("part_map", part_map),
             ]
             + list(SYMPY_INTERP.items())
         )
         closure_vars.update(CLOSURE_VARS)
+
+        # Let's go over how this code works.
+        #
+        # 1) The cache entry structure in eval_frame.c is stored in the frame's extra field. Each cache_entry is a node
+        # in a linked list. Each cache entry represents a compiled frame with specializations. In order to know if
+        # a cache entry's compiled frame is valid for reuse, we need to invoke a function, check_fn,
+        # to compare current state against the specializations captured at compile time.
+        #
+        # 2) The function, check_fn, mentioned above, is defined by executing the function below, ___make_guard_fn
+        #
+        # 3) In this, this code is rather confusing, because it defines both ___make_guard_fn and the lambda it produces
+        # which becomes the check_fn.
+        #
+        # 4) Everything is `code` becomes the check_fn. We write it out by first defining a `__fail`, which
+        # given a code_part_id, the id() of a code_part, extract a code_part from the part_map and binds a scope
+        # to it. This is used later for evaluating the expression defined in code_part, if necessary. `__fail` is
+        # invoked if a specific sub expression of a guard is failed.
+        #
+        # 5) The sub expressions of guards are defined 1 per code_part. We iterate over the code_parts and produce
+        # code that (a) assigns the result of the expression to a variable named `passing` (b) checks if not passing,
+        # (c) and if not passing, returns __fail(code_part_id) with the id of the code_part used to produce the code.
+        # or (d) if passing, proceeds until we've run all the sub expressions through.
+        #
+        # 6) In the event that we re-enter frame evaluation having failed a guard, we return the code_part
+        # and pass it through to the frame evaluation callback. This is where downstream systems that handle guard
+        # failures hook in, like guard failure logging, or converting static shape failures to dynamic shapes (if
+        # the config is set).
+        code = "      def __fail(code_part_id): \n"
+        code += "          code_part = part_map[code_part_id] \n"
+        code += "          code_part.scope = locals() \n"
+        code += "          code_part.scope['L'] = L \n"
```

Comment:
Yeah, so when you told me you were doing this inline, I was skeptical, and looking at the code here I am still skeptical. There are two reasons:

1. The performance of the fail case in guards is important when there are multiple cache entries and in the steady state we are using the cache to dispatch between several different possibilities. When this occurs, we will run the `__fail` code for every entry that comes before the eventual hit. So you don't want to be doing extra work in this case.
2. Arguably, the way to fix the performance above is setting up a tree structure of shared guards so that equivalent guards can be amortized across multiple cache entries. This would entail a substantial refactor of the guard logic here, in a way that would be pretty different from the way you've set things up here. So I'm still not a fan of pushing this code in that direction still...

Reply:
And no, I don't expect this to show up in our benchmark suite, because we're only running the model once on these tests!

Reply:
I suppose you might end up exercising the code here in case we run on a benchmark model with a lot of graph breaks, and on some of the inner frames that are retraced, we end up with multiple cache entries. Unfortunately, we don't track this statistic in the CI so it's hard to know how many models are like this.

Reply:
This is still really really cheap. Also, it allows you to avoid having to re-run all the failures in guard_fail_fn.

Reply:
What if we just executed the objects?

```
def guard_fn(args):
      for code_part in code_parts:
          if not code_part(args):
               return False, code_part
     return True, None
```

Reply:
> What if we just executed the objects?

I don't know what the perf impact of unrolling a loop in Python bytecode is. But two levels of interpreter (CPython interpreter + code_parts interpreter) seems worse than one level.

> This is still really really cheap. Also, it allows you to avoid having to re-run all the failures in guard_fail_fn.

How do you know it's really really cheap? More importantly, how do you know people aren't going to add more stuff to this in the future that will make it not cheap? (Case study: wasn't `guard_failure_fn` supposed to be cheap? But it actually turned into 1-2% overhead, as this PR shows?)

This being said, I think I will be willing to accept this code as is, as long as we have agreement on where the code is headed in the near future (since we're going to have to rework this anyway for CSE, and then rework it again for guard tree structure.)

Reply:
Yes, agreed below.